In [1]:
import pandas as pd
import numpy as np

In [2]:
pew_df = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\pew.csv')

### Tidying Data #1

In [3]:
print(pew_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   religion            18 non-null     object
 1   <$10k               18 non-null     int64 
 2   $10-20k             18 non-null     int64 
 3   $20-30k             18 non-null     int64 
 4   $30-40k             18 non-null     int64 
 5   $40-50k             18 non-null     int64 
 6   $50-75k             18 non-null     int64 
 7   $75-100k            18 non-null     int64 
 8   $100-150k           18 non-null     int64 
 9   >150k               18 non-null     int64 
 10  Don't know/refused  18 non-null     int64 
dtypes: int64(10), object(1)
memory usage: 1.7+ KB
None


In [4]:
pew_df.head()

religion  <$10k  $10-20k  $20-30k  $30-40k  $40-50k  $50-75k  \
0            Agnostic     27       34       60       81       76      137   
1             Atheist     12       27       37       52       35       70   
2            Buddhist     27       21       30       34       33       58   
3            Catholic    418      617      732      670      638     1116   
4  Don’t know/refused     15       14       15       11       10       35   

   $75-100k  $100-150k  >150k  Don't know/refused  
0       122        109     84                  96  
1        73         59     74                  76  
2        62         39     53                  54  
3       949        792    633                1489  
4        21         17     18                 116

Notice the the design of the table is inconvenient to navigate. I want to *unpivot* the table.

This is when `melt()` is used.

In [5]:
# id_vars are identifier columns
melted_pew = pew_df.melt(id_vars='religion', 
                         var_name='Income',
                         value_name='Count')

melted_pew.head(10)

religion Income  Count
0                 Agnostic  <$10k     27
1                  Atheist  <$10k     12
2                 Buddhist  <$10k     27
3                 Catholic  <$10k    418
4       Don’t know/refused  <$10k     15
5         Evangelical Prot  <$10k    575
6                    Hindu  <$10k      1
7  Historically Black Prot  <$10k    228
8        Jehovah's Witness  <$10k     20
9                   Jewish  <$10k     19

In [6]:
print(pew_df.shape)
print(melted_pew.shape)

(18, 11)
(180, 3)


Essentially, it is transforming wide format data into narrow format data by unpivoting values

### Tidying Data #2

In [7]:
billboard = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\billboard.csv')

In [8]:
print(billboard.shape)
print(billboard.columns)

(317, 81)
Index(['year', 'artist', 'track', 'time', 'date.entered', 'wk1', 'wk2', 'wk3',
       'wk4', 'wk5', 'wk6', 'wk7', 'wk8', 'wk9', 'wk10', 'wk11', 'wk12',
       'wk13', 'wk14', 'wk15', 'wk16', 'wk17', 'wk18', 'wk19', 'wk20', 'wk21',
       'wk22', 'wk23', 'wk24', 'wk25', 'wk26', 'wk27', 'wk28', 'wk29', 'wk30',
       'wk31', 'wk32', 'wk33', 'wk34', 'wk35', 'wk36', 'wk37', 'wk38', 'wk39',
       'wk40', 'wk41', 'wk42', 'wk43', 'wk44', 'wk45', 'wk46', 'wk47', 'wk48',
       'wk49', 'wk50', 'wk51', 'wk52', 'wk53', 'wk54', 'wk55', 'wk56', 'wk57',
       'wk58', 'wk59', 'wk60', 'wk61', 'wk62', 'wk63', 'wk64', 'wk65', 'wk66',
       'wk67', 'wk68', 'wk69', 'wk70', 'wk71', 'wk72', 'wk73', 'wk74', 'wk75',
       'wk76'],
      dtype='object')


In [9]:
melted_billboard = billboard.melt(id_vars=['year', 'artist', 'track', 'time', 'date.entered'],
                                  var_name='Week',
                                  value_name='Rating')
melted_billboard.head()

year        artist                    track  time date.entered Week  Rating
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1    91.0
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08  wk1    81.0
3  2000  3 Doors Down                    Loser  4:24   2000-10-21  wk1    76.0
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1    57.0

Let's pick an artist and a track to see if everything is formatted properly

In [10]:
melted_billboard.loc[(melted_billboard['artist'] == '3 Doors Down')&
                     (melted_billboard['track'] == 'Loser')].tail()

year        artist  track  time date.entered  Week  Rating
22510  2000  3 Doors Down  Loser  4:24   2000-10-21  wk72     NaN
22827  2000  3 Doors Down  Loser  4:24   2000-10-21  wk73     NaN
23144  2000  3 Doors Down  Loser  4:24   2000-10-21  wk74     NaN
23461  2000  3 Doors Down  Loser  4:24   2000-10-21  wk75     NaN
23778  2000  3 Doors Down  Loser  4:24   2000-10-21  wk76     NaN

In [11]:
print(melted_billboard.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24092 entries, 0 to 24091
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   year          24092 non-null  int64  
 1   artist        24092 non-null  object 
 2   track         24092 non-null  object 
 3   time          24092 non-null  object 
 4   date.entered  24092 non-null  object 
 5   Week          24092 non-null  object 
 6   Rating        5307 non-null   float64
dtypes: float64(1), int64(1), object(5)
memory usage: 1.3+ MB
None


Noticeable changes 
* Dramatic change of the shape (317, 81) -> (24092, 7)
* Memory usage went up from `200.7+ KB` to `1.3+ MB`
* There are numerous rows with nontype value for the `Rating` column

If I were to unpivot a much bigger dataset, memory usage might be a problem.

To reduce it, let's only select the rows that hold values for the `Rating` column

In [12]:
melted_billboard = melted_billboard[pd.notnull(melted_billboard['Rating'])]

In [13]:
melted_billboard.loc[(melted_billboard['artist'] == '3 Doors Down')&
                     (melted_billboard['track'] == 'Loser')].tail(10)

year        artist  track  time date.entered  Week  Rating
3173  2000  3 Doors Down  Loser  4:24   2000-10-21  wk11    61.0
3490  2000  3 Doors Down  Loser  4:24   2000-10-21  wk12    59.0
3807  2000  3 Doors Down  Loser  4:24   2000-10-21  wk13    61.0
4124  2000  3 Doors Down  Loser  4:24   2000-10-21  wk14    66.0
4441  2000  3 Doors Down  Loser  4:24   2000-10-21  wk15    72.0
4758  2000  3 Doors Down  Loser  4:24   2000-10-21  wk16    76.0
5075  2000  3 Doors Down  Loser  4:24   2000-10-21  wk17    75.0
5392  2000  3 Doors Down  Loser  4:24   2000-10-21  wk18    67.0
5709  2000  3 Doors Down  Loser  4:24   2000-10-21  wk19    73.0
6026  2000  3 Doors Down  Loser  4:24   2000-10-21  wk20    70.0

In [14]:
print(melted_billboard.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5307 entries, 0 to 20350
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   year          5307 non-null   int64  
 1   artist        5307 non-null   object 
 2   track         5307 non-null   object 
 3   time          5307 non-null   object 
 4   date.entered  5307 non-null   object 
 5   Week          5307 non-null   object 
 6   Rating        5307 non-null   float64
dtypes: float64(1), int64(1), object(5)
memory usage: 331.7+ KB
None


### Tidying Data #3

In [15]:
weather = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\weather.csv')

In [16]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 35 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       22 non-null     object 
 1   year     22 non-null     int64  
 2   month    22 non-null     int64  
 3   element  22 non-null     object 
 4   d1       2 non-null      float64
 5   d2       4 non-null      float64
 6   d3       4 non-null      float64
 7   d4       2 non-null      float64
 8   d5       8 non-null      float64
 9   d6       2 non-null      float64
 10  d7       2 non-null      float64
 11  d8       2 non-null      float64
 12  d9       0 non-null      float64
 13  d10      2 non-null      float64
 14  d11      2 non-null      float64
 15  d12      0 non-null      float64
 16  d13      2 non-null      float64
 17  d14      4 non-null      float64
 18  d15      2 non-null      float64
 19  d16      2 non-null      float64
 20  d17      2 non-null      float64
 21  d18      0 non-nul

In [17]:
weather.head()

id  year  month element  d1    d2    d3  d4    d5  d6  ...  d22   d23  \
0  MX17004  2010      1    tmax NaN   NaN   NaN NaN   NaN NaN  ...  NaN   NaN   
1  MX17004  2010      1    tmin NaN   NaN   NaN NaN   NaN NaN  ...  NaN   NaN   
2  MX17004  2010      2    tmax NaN  27.3  24.1 NaN   NaN NaN  ...  NaN  29.9   
3  MX17004  2010      2    tmin NaN  14.4  14.4 NaN   NaN NaN  ...  NaN  10.7   
4  MX17004  2010      3    tmax NaN   NaN   NaN NaN  32.1 NaN  ...  NaN   NaN   

   d24  d25  d26  d27  d28  d29   d30  d31  
0  NaN  NaN  NaN  NaN  NaN  NaN  27.8  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  14.5  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN   NaN  NaN  

[5 rows x 35 columns]

In [18]:
melted_weather = weather.melt(id_vars=['id', 'year', 'month', 'element'],
                              value_name='temperature',
                              var_name='day')
melted_weather.head()

id  year  month element day  temperature
0  MX17004  2010      1    tmax  d1          NaN
1  MX17004  2010      1    tmin  d1          NaN
2  MX17004  2010      2    tmax  d1          NaN
3  MX17004  2010      2    tmin  d1          NaN
4  MX17004  2010      3    tmax  d1          NaN

In [19]:
melted_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 682 entries, 0 to 681
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           682 non-null    object 
 1   year         682 non-null    int64  
 2   month        682 non-null    int64  
 3   element      682 non-null    object 
 4   day          682 non-null    object 
 5   temperature  66 non-null     float64
dtypes: float64(1), int64(2), object(3)
memory usage: 32.1+ KB


It looks a lot easier to comprehend than before, but some minor changes are required.

The better design would be having the max & min temperatures as separate columns to avoid having too many rows.

In [20]:
# store tmax and tmin values separately
tmax = melted_weather.loc[melted_weather['element'] == 'tmax']['temperature']
tmin = melted_weather.loc[melted_weather['element'] == 'tmin']['temperature']

# drop element and temperature columns
# drop duplicates so there are only unique values for days
melted_weather.drop(columns=['element', 'temperature'], inplace=True)
melted_weather.drop_duplicates(inplace=True)

# adding columns & resetting indexes
melted_weather['max_temperature'] = tmax
melted_weather['min_temperature'] = tmin
melted_weather.reset_index(drop=True,inplace=True)
melted_weather

id  year  month  day  max_temperature  min_temperature
0    MX17004  2010      1   d1              NaN              NaN
1    MX17004  2010      2   d1              NaN              NaN
2    MX17004  2010      3   d1              NaN              NaN
3    MX17004  2010      4   d1              NaN              NaN
4    MX17004  2010      5   d1              NaN              NaN
..       ...   ...    ...  ...              ...              ...
336  MX17004  2010      7  d31              NaN              NaN
337  MX17004  2010      8  d31             25.4              NaN
338  MX17004  2010     10  d31              NaN              NaN
339  MX17004  2010     11  d31              NaN              NaN
340  MX17004  2010     12  d31              NaN              NaN

[341 rows x 6 columns]

In [21]:
# checking if there is any more duplicate
print(all(melted_weather.duplicated()))

False


In [22]:
melted_weather.isnull().sum()

id                   0
year                 0
month                0
day                  0
max_temperature    308
min_temperature    341
dtype: int64

In [23]:
melted_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341 entries, 0 to 340
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               341 non-null    object 
 1   year             341 non-null    int64  
 2   month            341 non-null    int64  
 3   day              341 non-null    object 
 4   max_temperature  33 non-null     float64
 5   min_temperature  0 non-null      float64
dtypes: float64(2), int64(2), object(2)
memory usage: 16.1+ KB


It's too bad 308 of 341 rows have nontype values.

Other than that, everything looks well organized and tidy.


shape changes: (22, 35) -> (682, 6) -> (341, 6)

memory changes: 6.1+ KB -> 32.1+ KB -> 16.1+ KB

### Tidying Data #4

In [24]:
country_df = pd.read_csv(r'C:\Users\taewoo\PycharmProjects\data_professionals\Projects\Datasets\country_timeseries.csv')

In [25]:
print(country_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122 entries, 0 to 121
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date                 122 non-null    object 
 1   Day                  122 non-null    int64  
 2   Cases_Guinea         93 non-null     float64
 3   Cases_Liberia        83 non-null     float64
 4   Cases_SierraLeone    87 non-null     float64
 5   Cases_Nigeria        38 non-null     float64
 6   Cases_Senegal        25 non-null     float64
 7   Cases_UnitedStates   18 non-null     float64
 8   Cases_Spain          16 non-null     float64
 9   Cases_Mali           12 non-null     float64
 10  Deaths_Guinea        92 non-null     float64
 11  Deaths_Liberia       81 non-null     float64
 12  Deaths_SierraLeone   87 non-null     float64
 13  Deaths_Nigeria       38 non-null     float64
 14  Deaths_Senegal       22 non-null     float64
 15  Deaths_UnitedStates  18 non-null     flo

Instead of having all these separate colums for deaths and cases, I can store everything in two columns: `Cases` and `Deaths`.

In [26]:
country_df.head()

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0    1/5/2015  289        2776.0            NaN            10030.0   
1    1/4/2015  288        2775.0            NaN             9780.0   
2    1/3/2015  287        2769.0         8166.0             9722.0   
3    1/2/2015  286           NaN         8157.0                NaN   
4  12/31/2014  284        2730.0         8115.0             9633.0   

   Cases_Nigeria  Cases_Senegal  Cases_UnitedStates  Cases_Spain  Cases_Mali  \
0            NaN            NaN                 NaN          NaN         NaN   
1            NaN            NaN                 NaN          NaN         NaN   
2            NaN            NaN                 NaN          NaN         NaN   
3            NaN            NaN                 NaN          NaN         NaN   
4            NaN            NaN                 NaN          NaN         NaN   

   Deaths_Guinea  Deaths_Liberia  Deaths_SierraLeone  Deaths_Nigeria  \
0         1786.0             NaN              2977.0             NaN   
1         1781.0             NaN              2943.0             NaN   
2         1767.0          3496.0              2915.0             NaN   
3            NaN          3496.0                 NaN             NaN   
4         1739.0          3471.0              2827.0             NaN   

   Deaths_Senegal  Deaths_UnitedStates  Deaths_Spain  Deaths_Mali  
0             NaN                  NaN           NaN          NaN  
1             NaN                  NaN           NaN          NaN  
2             NaN                  NaN           NaN          NaN  
3             NaN                  NaN           NaN          NaN  
4             NaN                  NaN           NaN          NaN

In [27]:
melted_country = country_df.melt(id_vars=['Date', 'Day'],
                                 value_name='Amount')
melted_country.loc[melted_country['Date'] == '1/5/2015']
melted_country[['Type', 'Country']] = melted_country['variable'].str.split('_', expand=True)
melted_country.drop(columns='variable', inplace=True)
melted_country = melted_country[['Date', 'Day', 'Country', 'Type', 'Amount']]

In [28]:
melted_country

Date  Day Country    Type  Amount
0       1/5/2015  289  Guinea   Cases  2776.0
1       1/4/2015  288  Guinea   Cases  2775.0
2       1/3/2015  287  Guinea   Cases  2769.0
3       1/2/2015  286  Guinea   Cases     NaN
4     12/31/2014  284  Guinea   Cases  2730.0
...          ...  ...     ...     ...     ...
1947   3/27/2014    5    Mali  Deaths     NaN
1948   3/26/2014    4    Mali  Deaths     NaN
1949   3/25/2014    3    Mali  Deaths     NaN
1950   3/24/2014    2    Mali  Deaths     NaN
1951   3/22/2014    0    Mali  Deaths     NaN

[1952 rows x 5 columns]

In [29]:
print(melted_country.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1952 entries, 0 to 1951
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Date     1952 non-null   object 
 1   Day      1952 non-null   int64  
 2   Country  1952 non-null   object 
 3   Type     1952 non-null   object 
 4   Amount   738 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 76.4+ KB
None


There are both `Cases` and `Deaths` in the column named `Type`.

In order to alter them as columns, I need temporary dataframes to store `Cases` and `Deaths`

In [30]:
# I have to split cases and deaths, so used loc to create separate dataframes
cases_df = melted_country.copy().loc[melted_country['Type']=='Cases']
cases_df.drop(columns='Type', inplace=True)
cases_df.rename(columns={'Amount':'Cases'}, inplace=True)
deaths_df = melted_country.copy().loc[melted_country['Type']=='Deaths']
deaths_df.drop(columns='Type', inplace=True)
deaths_df.rename(columns={'Amount':'Deaths'}, inplace=True)

# join two tables on date, country, and day
tidy_country = cases_df.merge(deaths_df, on=['Date', 'Country', 'Day'])

In [31]:
tidy_country.head(10)

Date  Day Country   Cases  Deaths
0    1/5/2015  289  Guinea  2776.0  1786.0
1    1/4/2015  288  Guinea  2775.0  1781.0
2    1/3/2015  287  Guinea  2769.0  1767.0
3    1/2/2015  286  Guinea     NaN     NaN
4  12/31/2014  284  Guinea  2730.0  1739.0
5  12/28/2014  281  Guinea  2706.0  1708.0
6  12/27/2014  280  Guinea  2695.0  1697.0
7  12/24/2014  277  Guinea  2630.0     NaN
8  12/21/2014  273  Guinea  2597.0  1607.0
9  12/20/2014  272  Guinea  2571.0  1586.0

In [32]:
tidy_country.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 976 entries, 0 to 975
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Date     976 non-null    object 
 1   Day      976 non-null    int64  
 2   Country  976 non-null    object 
 3   Cases    372 non-null    float64
 4   Deaths   366 non-null    float64
dtypes: float64(2), int64(1), object(2)
memory usage: 45.8+ KB


Compared to the original dataframe, it looks much compact and organized.

I was able to change the shape from (122, 18) to (976, 5) 